### PR TITLE
[Merged by Bors] - feat: port Data.Finset.Sups

### PR DESCRIPTION
**Pull request by mo271**
- [x] depends on: https://github.com/leanprover-community/mathlib/pull/18172
 - [x]  depends on: #2991 
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Associated.lean [hunk 1]
```diff
@@ -266,7 +266,7 @@ theorem irreducible_isUnit_mul {a b : α} (h : IsUnit a) : Irreducible (a * b) 
 
 theorem irreducible_mul_units (a : αˣ) (b : α) : Irreducible (b * ↑a) ↔ Irreducible b := by
   simp only [irreducible_iff, Units.isUnit_mul_units, and_congr_right_iff]
-  refine' fun _ => ⟨fun h A B HAB => _, fun h A B HAB => _⟩
+ h A B HAB => _, fun h A B HAB => _⟩
```
Comment (urkud) [inline Mathlib/Algebra/Associated.lean:269]: Something strange happens with indentation here.

Comment (mo271) [reply Mathlib/Algebra/Associated.lean:269]: WIP...

Comment (mo271) [reply Mathlib/Algebra/Associated.lean]: thanks, removed this spurious edit..

File: Mathlib/Data/Finset/Sups.lean [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+  card_image₂_iff
+#align finset.card_sups_iff Finset.cardSups_iff
+
+variable {s s₁ s₂ t t₁ t₂ u}
+
+theorem sup_memSups : a ∈ s → b ∈ t → a ⊔ b ∈ s ⊻ t :=
+  mem_image₂_of_mem
+#align finset.sup_mem_sups Finset.sup_memSups
+
+theorem sups_subset : s₁ ⊆ s₂ → t₁ ⊆ t₂ → s₁ ⊻ t₁ ⊆ s₂ ⊻ t₂ :=
```
Comment (urkud) [inline Mathlib/Data/Finset/Sups.lean:88]: Should it be `@[mono]` (both here and in `mathlib3`)?

File: Mathlib/Data/Finset/Sups.lean [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+
+theorem sups_subset_left : t₁ ⊆ t₂ → s ⊻ t₁ ⊆ s ⊻ t₂ :=
+  image₂_subset_left
+#align finset.sups_subset_left Finset.sups_subset_left
+
+theorem sups_subset_right : s₁ ⊆ s₂ → s₁ ⊻ t ⊆ s₂ ⊻ t :=
+  image₂_subset_right
+#align finset.sups_subset_right Finset.sups_subset_right
+
+theorem image_subsetSups_left : b ∈ t → (fun a => a ⊔ b) '' s ⊆ s ⊻ t :=
```
Comment (urkud) [inline Mathlib/Data/Finset/Sups.lean:100]:
```suggestion
theorem image_subsetSups_left : b ∈ t → (· ⊔ b) '' s ⊆ s ⊻ t :=
```

File: Mathlib/Data/Finset/Sups.lean [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+  image₂_empty_right
+#align finset.sups_empty_right Finset.sups_empty_right
+
+@[simp]
+theorem sups_eq_empty_iff : s ⊻ t = ∅ ↔ s = ∅ ∨ t = ∅ :=
+  image₂_eq_empty_iff
+#align finset.sups_eq_empty_iff Finset.sups_eq_empty_iff
+
+@[simp]
+theorem sups_singleton_left : {a} ⊻ t = t.image fun b => a ⊔ b :=
```
Comment (urkud) [inline Mathlib/Data/Finset/Sups.lean:150]:
```suggestion
theorem sups_singleton_left : {a} ⊻ t = t.image (a ⊔ ·) :=
```

File: Mathlib/Data/Finset/Sups.lean [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+  image₂_singleton_left
+#align finset.sups_singleton_left Finset.sups_singleton_left
+
+@[simp]
+theorem sups_singleton_right : s ⊻ {b} = s.image fun a => a ⊔ b :=
```
Comment (urkud) [inline Mathlib/Data/Finset/Sups.lean:155]:
```suggestion
theorem sups_singleton_right : s ⊻ {b} = s.image (· ⊔ b) :=
```

File: Mathlib/Data/Finset/Sups.lean [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+  image₂_singleton_right
+#align finset.sups_singleton_right Finset.sups_singleton_right
+
+theorem sups_singleton_left' : {a} ⊻ t = t.image ((· ⊔ ·) a) :=
```
Comment (urkud) [inline Mathlib/Data/Finset/Sups.lean:159]: Do we need this version in Mathlib 4?

File: Mathlib/Data/Finset/Sups.lean (hunk 2 above)
Comment (mo271) [reply Mathlib/Data/Finset/Sups.lean:88]: Don't know, did @YaelDillies write that code?

**Comment by mo271** [comment]
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Data.2EFinset.2ESups.20mathlib4.231663

**Comment by mo271** [comment]
Things left to do:

- [ ] fix disjSups_comm
- [ ] fix two `redundant binder annotation update` error
- [ ] decide if `sups_singleton_left'` is needed

**Comment by YaelDillies** [comment]
This file is actively being worked on. Please wait for https://github.com/leanprover-community/mathlib/pull/18172

File: Mathlib/Data/Finset/Sups.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Data/Finset/Sups.lean:88]: Possibly, yeah

**Comment by mo271** [comment]
> This file is actively being worked on. Please wait for [leanprover-community/mathlib#18172](https://github.com/leanprover-community/mathlib/pull/18172)

Sure. (I didn't know, there was nothing in the porting note, it just showed up there as a file that needed porting...)

**Comment by YaelDillies** [comment]
Yeah sorry, I had no idea it was up for porting.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib#18172
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib#18172~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by mo271** [comment]
Sounds goodAm 18.03.2023 12:14 schrieb Yaël Dillies ***@***.***>:
There are now enough changes to the mathlib file that I think it'll be easier if you start over.

**Comment by mo271** [comment]
Apparently there is a new dependency, namely Data.Set.Sups, #2991

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib#18172~~
* ~~leanprover-community/mathlib4#2991~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by mo271** [comment]
I rebased and started over. Three things left, and also decide how many of the names shoudl be `fooSups` instead of `foo_sups`

**Comment by mo271** [comment]
TODO:

- [ ] `image_subset_sups_right`
- [ ] `image_subset_infs_right`
- [ ] `disjSups_comm`
- [ ] naming?

The `disjSups_comm` issue is the same problem as described above.


For `image_subset_sups_right` and `image_subset_infs_right`, it is not clear to me why this is not working. It wasn't an issue before...

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**